src/driver/main.cpp

```diff
@@ -0,0 +1,214 @@
+    }
+
+    program load()
+    {
+        program p;
+        if(type.empty())
+        {
+            if(ends_with(file, ".onnx"))
+                type = "onnx";
+            else
+                type = "tf";
+        }
+        std::cout << "Reading: " << file << std::endl;
+        if(type == "onnx")
+            p = parse_onnx(file);
+        else if(type == "tf")
+            p = parse_tf(file, is_nhwc);
+        if(trim > 0)
+        {
+            auto last = std::prev(p.end(), trim);
+            p.remove_instructions(last, p.end());
+        }
+        return p;
+    }
+};
+
+struct compiler
+{
+    loader l;
+    bool gpu = true;
+    void parse(argument_parser& ap)
+    {
+        l.parse(ap);
+        ap(gpu, {"--gpu"}, ap.help("Compile on the gpu"), ap.set_value(true));
+        ap(gpu, {"--cpu"}, ap.help("Compile on the cpu"), ap.set_value(false));
+    }
+
+    program compile()
+    {
+        auto p = l.load();
+        compile_program(p, gpu);
+        return p;
+    }
+
+    auto params(const program& p) { return create_param_map(p, gpu); }
+};
+
+struct read : command<read>
+{
+    loader l;
+    void parse(argument_parser& ap) { l.parse(ap); }
+
+    void run()
+    {
+        auto p = l.load();
+        std::cout << p << std::endl;
+    }
+};
+
+struct verify : command<verify>
+{
+    loader l;
+    double tolerance     = 80;
+    bool per_instruction = false;
+    bool reduce          = false;
+    void parse(argument_parser& ap)
+    {
+        l.parse(ap);
+        ap(tolerance, {"--tolerance"}, ap.help("Tolerance for errors"));
+        ap(per_instruction,
+           {"-i", "--per-instruction"},
+           ap.help("Verify each instruction"),
+           ap.set_value(true));
+        ap(reduce, {"-r", "--reduce"}, ap.help("Reduce program and verify"), ap.set_value(true));
+    }
+
+    void run()
+    {
+        auto p = l.load();
+        std::cout << p << std::endl;
+
+        if(per_instruction)
+        {
+            verify_instructions(p, tolerance);
+        }
+        else if(reduce)
+        {
+            verify_reduced_program(p, tolerance);
+        }
+        else
+        {
+            verify_program(l.file, p, tolerance);
+        }
+    }
+};
+
+struct compile : command<compile>
+{
+    compiler c;
+    void parse(argument_parser& ap) { c.parse(ap); }
+
+    void run()
+    {
+        std::cout << "Compiling ... " << std::endl;
+        auto p = c.compile();
+        std::cout << p << std::endl;
+    }
+};
+
+struct run_cmd : command<run_cmd>
+{
+    compiler c;
+    void parse(argument_parser& ap) { c.parse(ap); }
+
+    void run()
+    {
+        std::cout << "Compiling ... " << std::endl;
+        auto p = c.compile();
+        std::cout << "Allocating params ... " << std::endl;
+        auto m = c.params(p);
+        p.eval(m);
+        std::cout << p << std::endl;
+    }
+};
+
+struct perf : command<perf>
+{
+    compiler c;
+    unsigned n = 100;
```

Comment:
Might be useful to print out what default value of n is set to.

Reply:
It prints out the default now with `--help` flag.